REGRESSION(276009@main): Crash in ApplicationStateTracker::applicationDidEnterBackground()

jernoble opened the pull request:
#### 15b191da20fcceee30cd9cb3b1e98c77ffe2b565
<pre>
REGRESSION(276009@main): Crash in ApplicationStateTracker::applicationDidEnterBackground()
<a href="https://bugs.webkit.org/show_bug.cgi?id=272241">https://bugs.webkit.org/show_bug.cgi?id=272241</a>
<a href="https://rdar.apple.com/125895046">rdar://125895046</a>

Reviewed by Eric Carlson.

From crash reports, it appears that the NSNotificationCenter observers are being called after
the ApplicationStateTracker itself has been destroyed. Ensure these observers are all
unregistered upon destruction, and additionally protect against the observers being called
after unregistration by passing a WeakPtr into the observer&apos;s lambda.

* Source/WebKit/UIProcess/ApplicationStateTracker.h:
* Source/WebKit/UIProcess/ApplicationStateTracker.mm:
(WebKit::ApplicationStateTracker::~ApplicationStateTracker):
(WebKit::ApplicationStateTracker::removeAllObservers):
(WebKit::ApplicationStateTracker::setScene):
(WebKit::ApplicationStateTracker::setViewController):

Canonical link: <a href="https://commits.webkit.org/277170@main">https://commits.webkit.org/277170@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cad99740df2a693d9b67f53b324903481a3ff376

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/46803 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/25964 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/49426 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/49480 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/42850 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/30368 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/23430 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/38125 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/47383 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/22960 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/40308 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/19423 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/20289 "Passed tests") | | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/4849 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/43042 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/41817 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/51353 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/21812 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/18161 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/45410 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/23100 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/44370 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10353 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/23603 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/22809 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/UIProcess/ApplicationStateTracker.mm:
```diff
@@ -167,10 +167,33 @@ ApplicationType applicationType(UIWindow *window)
 
+void ApplicationStateTracker::removeAllObservers()
+{
+    NSNotificationCenter *notificationCenter = [NSNotificationCenter defaultCenter];
```
We should use a RetainPtr.

jernoble replied:
`defaultCenter` is a singleton object; it's guaranteed to be the same object every time `+defaultCenter` is called, so no retain is needed.

cdumez replied:
Should I point you to the smart pointer guidelines?

cdumez replied:
The times of "trust me it's OK in this case" are long gone. We have a set of smart pointers that we apply consistently, no matter what.

cdumez commented on Source/WebKit/UIProcess/ApplicationStateTracker.mm:
```diff
@@ -262,21 +278,30 @@ ApplicationType applicationType(UIWindow *window)
     RELEASE_LOG(ViewState, "%p - ApplicationStateTracker::ApplicationStateTracker(): m_isInBackground=%d", this, m_isInBackground);
 
-    m_didEnterBackgroundObserver = [notificationCenter addObserverForName:UISceneDidEnterBackgroundNotification object:scene queue:nil usingBlock:[this](NSNotification *notification) {
+    NSNotificationCenter *notificationCenter = [NSNotificationCenter defaultCenter];
```
RetainPtr.

cdumez commented on Source/WebKit/UIProcess/ApplicationStateTracker.mm:
```diff
@@ -262,21 +278,30 @@ ApplicationType applicationType(UIWindow *window)
 
-    m_didEnterBackgroundObserver = [notificationCenter addObserverForName:UISceneDidEnterBackgroundNotification object:scene queue:nil usingBlock:[this](NSNotification *notification) {
+    NSNotificationCenter *notificationCenter = [NSNotificationCenter defaultCenter];
+    m_didEnterBackgroundObserver = [notificationCenter addObserverForName:UISceneDidEnterBackgroundNotification object:scene queue:nil usingBlock:[this, weakThis = WeakPtr { this }](NSNotification *notification) {
```
`= WeakThis { *this }` to avoid the null check.

cdumez commented on Source/WebKit/UIProcess/ApplicationStateTracker.mm:
```diff
@@ -262,21 +278,30 @@ ApplicationType applicationType(UIWindow *window)
     }];
 
-    m_willEnterForegroundObserver = [notificationCenter addObserverForName:UISceneWillEnterForegroundNotification object:scene queue:nil usingBlock:[this](NSNotification *notification) {
+    m_willEnterForegroundObserver = [notificationCenter addObserverForName:UISceneWillEnterForegroundNotification object:scene queue:nil usingBlock:[this, weakThis = WeakPtr { this }](NSNotification *notification) {
```
`WeakThis { *this }`

cdumez commented on Source/WebKit/UIProcess/ApplicationStateTracker.mm:
```diff
@@ -262,21 +278,30 @@ ApplicationType applicationType(UIWindow *window)
     }];
 
-    m_willBeginSnapshotSequenceObserver = [notificationCenter addObserverForName:_UISceneWillBeginSystemSnapshotSequence object:scene queue:nil usingBlock:[this](NSNotification *notification) {
+    m_willBeginSnapshotSequenceObserver = [notificationCenter addObserverForName:_UISceneWillBeginSystemSnapshotSequence object:scene queue:nil usingBlock:[this, weakThis = WeakPtr { this }](NSNotification *notification) {
```
ditto.

cdumez commented on Source/WebKit/UIProcess/ApplicationStateTracker.mm:
```diff
@@ -262,21 +278,30 @@ ApplicationType applicationType(UIWindow *window)
     }];
 
-    m_didCompleteSnapshotSequenceObserver = [notificationCenter addObserverForName:_UISceneDidCompleteSystemSnapshotSequence object:scene queue:nil usingBlock:[this](NSNotification *notification) {
+    m_didCompleteSnapshotSequenceObserver = [notificationCenter addObserverForName:_UISceneDidCompleteSystemSnapshotSequence object:scene queue:nil usingBlock:[this, weakThis = WeakPtr { this }](NSNotification *notification) {
```
ditto.

cdumez commented on Source/WebKit/UIProcess/ApplicationStateTracker.mm:
```diff
@@ -312,11 +332,16 @@ ApplicationType applicationType(UIWindow *window)
 
-    m_didEnterBackgroundObserver = [notificationCenter addObserverForName:viewServiceBackgroundNotificationName object:serviceViewController queue:nil usingBlock:[this, applicationPID](NSNotification *) {
+    NSNotificationCenter *notificationCenter = [NSNotificationCenter defaultCenter];
+    m_didEnterBackgroundObserver = [notificationCenter addObserverForName:viewServiceBackgroundNotificationName object:serviceViewController queue:nil usingBlock:[this, weakThis = WeakPtr { this }, applicationPID](NSNotification *) {
```
ditto.

cdumez commented on Source/WebKit/UIProcess/ApplicationStateTracker.mm:
```diff
@@ -312,11 +332,16 @@ ApplicationType applicationType(UIWindow *window)
     RELEASE_LOG(ProcessSuspension, "%{public}s has PID %d, host application PID=%d, isInBackground=%d", _UIApplicationIsExtension() ? "Extension" : "ViewService", getpid(), applicationPID, m_isInBackground);
 
-    m_didEnterBackgroundObserver = [notificationCenter addObserverForName:viewServiceBackgroundNotificationName object:serviceViewController queue:nil usingBlock:[this, applicationPID](NSNotification *) {
+    NSNotificationCenter *notificationCenter = [NSNotificationCenter defaultCenter];
```
RetainPtr.

cdumez commented on Source/WebKit/UIProcess/ApplicationStateTracker.mm:
```diff
@@ -312,11 +332,16 @@ ApplicationType applicationType(UIWindow *window)
         applicationDidEnterBackground();
     }];
-    m_willEnterForegroundObserver = [notificationCenter addObserverForName:viewServiceForegroundNotificationName object:serviceViewController queue:nil usingBlock:[this, applicationPID](NSNotification *) {
+    m_willEnterForegroundObserver = [notificationCenter addObserverForName:viewServiceForegroundNotificationName object:serviceViewController queue:nil usingBlock:[this, weakThis = WeakPtr { this }, applicationPID](NSNotification *) {
```
`WeakPtr { *this }`

jernoble committed:
REGRESSION(276009@main): Crash in ApplicationStateTracker::applicationDidEnterBackground()
https://bugs.webkit.org/show_bug.cgi?id=272241
rdar://125895046

Reviewed by Eric Carlson.

From crash reports, it appears that the NSNotificationCenter observers are being called after
the ApplicationStateTracker itself has been destroyed. Ensure these observers are all
unregistered upon destruction, and additionally protect against the observers being called
after unregistration by passing a WeakPtr into the observer's lambda.

* Source/WebKit/UIProcess/ApplicationStateTracker.h:
* Source/WebKit/UIProcess/ApplicationStateTracker.mm:
(WebKit::ApplicationStateTracker::~ApplicationStateTracker):
(WebKit::ApplicationStateTracker::removeAllObservers):
(WebKit::ApplicationStateTracker::setScene):
(WebKit::ApplicationStateTracker::setViewController):

Canonical link: https://commits.webkit.org/277170@main
